staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/objectmeta/validation.go

```diff
@@ -0,0 +1,123 @@
+	utilvalidation "k8s.io/apimachinery/pkg/util/validation"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+// Validate validates embedded ObjectMeta and TypeMeta.
+// It also validate those at the root if includeRoot is true.
+func Validate(pth *field.Path, obj interface{}, s *structuralschema.Structural, includeRoot bool) field.ErrorList {
+	if includeRoot {
+		if s == nil {
+			s = &structuralschema.Structural{}
+		}
+		clone := *s
+		clone.XEmbeddedResource = true
+		s = &clone
+	}
+	return validate(pth, obj, s)
+}
+
+func validate(pth *field.Path, x interface{}, s *structuralschema.Structural) field.ErrorList {
+	if s == nil {
+		return nil
+	}
+
+	var allErrs field.ErrorList
+
+	switch x := x.(type) {
+	case map[string]interface{}:
+		if s.XEmbeddedResource {
+			allErrs = append(allErrs, validateEmbeddedResource(pth, x, s)...)
+		}
+
+		for k, v := range x {
+			prop, ok := s.Properties[k]
+			if ok {
+				allErrs = append(allErrs, validate(pth.Child(k), v, &prop)...)
+			} else if s.AdditionalProperties != nil {
+				allErrs = append(allErrs, validate(pth.Key(k), v, s.AdditionalProperties.Structural)...)
+			}
+		}
+	case []interface{}:
+		for i, v := range x {
+			allErrs = append(allErrs, validate(pth.Index(i), v, s.Items)...)
+		}
+	default:
+		// scalars, do nothing
+	}
+
+	return allErrs
+}
+
+func validateEmbeddedResource(pth *field.Path, x map[string]interface{}, s *structuralschema.Structural) field.ErrorList {
+	var allErrs field.ErrorList
+
+	// require apiVersion and kind, but not metadata
+	if s.XEmbeddedResource {
+		if _, found := x["apiVersion"]; !found {
+			allErrs = append(allErrs, field.Required(pth.Child("apiVersion"), "must not be empty"))
+		}
+		if _, found := x["kind"]; !found {
+			allErrs = append(allErrs, field.Required(pth.Child("kind"), "must not be empty"))
+		}
+	}
+
+	for k, v := range x {
+		if s.XEmbeddedResource {
+			switch k {
+			case "apiVersion":
+				if apiVersion, ok := v.(string); !ok {
+					allErrs = append(allErrs, field.Invalid(pth.Child("apiVersion"), apiVersion, "must be a string"))
+				} else if len(apiVersion) == 0 {
+					allErrs = append(allErrs, field.Invalid(pth.Child("apiVersion"), apiVersion, "must not be empty"))
+				} else if _, err := schema.ParseGroupVersion(apiVersion); err != nil {
+					allErrs = append(allErrs, field.Invalid(pth.Child("apiVersion"), apiVersion, err.Error()))
+				}
+			case "kind":
+				if kind, ok := v.(string); !ok {
+					allErrs = append(allErrs, field.Invalid(pth.Child("kind"), v, "must be a string"))
+				} else if len(kind) == 0 {
+					allErrs = append(allErrs, field.Invalid(pth.Child("kind"), kind, "must not be empty"))
+				} else if errs := utilvalidation.IsDNS1035Label(strings.ToLower(kind)); len(errs) > 0 {
+					allErrs = append(allErrs, field.Invalid(pth.Child("kind"), kind, "may have mixed case, but should otherwise match: "+strings.Join(errs, ",")))
+				}
+			case "metadata":
+				meta, _, err := GetObjectMeta(x, true)
```

Comment:
Yeah, fixed.

The true or false here does not matter because we already fail during coercion. But changed this to false anyway as two layers of checks are safer.